core-rust/state-manager/src/protocol/protocol_update.rs

```diff
@@ -0,0 +1,489 @@
+pub trait ProtocolUpdater {
+    fn protocol_version_name(&self) -> String;
+    fn state_computer_configurator(&self) -> StateComputerConfigurator;
+    fn state_update_executor(&self) -> Box<dyn StateUpdateExecutor>;
+}
+
+#[derive(Clone, Debug)]
+pub struct StateComputerConfigurator {
+    pub network: NetworkDefinition,
+    pub logging_config: LoggingConfig,
+    pub validation_config: ValidationConfig,
+    pub costing_parameters: CostingParameters,
+}
+
+impl StateComputerConfigurator {
+    pub fn default(network: NetworkDefinition) -> StateComputerConfigurator {
+        let network_id = network.id;
+        StateComputerConfigurator {
+            network,
+            logging_config: LoggingConfig::default(),
+            validation_config: ValidationConfig::default(network_id),
+            costing_parameters: CostingParameters::default(),
+        }
+    }
+}
+
+impl StateComputerConfigurator {
+    pub fn ledger_transaction_validator(&self) -> LedgerTransactionValidator {
+        LedgerTransactionValidator::default_from_validation_config(self.validation_config)
+    }
+
+    pub fn user_transaction_validator(&self) -> NotarizedTransactionValidator {
+        NotarizedTransactionValidator::new(self.validation_config)
+    }
+
+    pub fn validation_config(&self) -> ValidationConfig {
+        self.validation_config
+    }
+
+    pub fn execution_configurator(&self, no_fees: bool) -> ExecutionConfigurator {
+        let mut costing_parameters = self.costing_parameters;
+        if no_fees {
+            costing_parameters.execution_cost_unit_price = Decimal::ZERO;
+            costing_parameters.finalization_cost_unit_price = Decimal::ZERO;
+            costing_parameters.state_storage_price = Decimal::ZERO;
+            costing_parameters.archive_storage_price = Decimal::ZERO;
+        }
+        ExecutionConfigurator::new(&self.network, &self.logging_config, costing_parameters)
+    }
+}
+
+pub trait StateUpdateExecutor {
+    /// Executes these state updates associated with the given protocol version
+    /// that haven't yet been applied
+    /// (hence "remaining", e.g. if node is restarted mid-protocol update).
+    fn execute_remaining_state_updates(&self);
+}
+
+/// A protocol updater implementation that only changes the configuration
+/// and does not commit any state updates.
+pub struct NoStateUpdatesProtocolUpdater {
+    protocol_version_name: String,
+    state_computer_configurator: StateComputerConfigurator,
+}
+
+impl NoStateUpdatesProtocolUpdater {
+    pub fn default(protocol_version_name: String, network: NetworkDefinition) -> Self {
+        Self {
+            protocol_version_name,
+            state_computer_configurator: StateComputerConfigurator::default(network),
+        }
+    }
+}
+
+impl ProtocolUpdater for NoStateUpdatesProtocolUpdater {
+    fn protocol_version_name(&self) -> String {
+        self.protocol_version_name.clone()
```

Comment:
(just a hunch) Whenever I see the same method impl in each struct implementing a trait, I suspect the responsibility should be moved one level higher.

Reply:
Yeah, I agree

Reply:
(removed)